Implementation of SugarSelection

talagayev opened the pull request:
Fixes #4563 

Changes made in this Pull Request:
 - Creation of the `SugarSelection` class that allows the selection of sugars through
 the access of known PDB, CHARMM and GLYCAM abbreviations
 - Addition of `GLYCAM` and `SUGAR_PDB` files in `MDAnalysisTest.data`
 - Addition of `test_sugar_glycam_selection()` and `test_sugar_pdb_selection`
 in `test_atomselections.py`

Currently I used the following abbreviations:

https://glycam.org/docs/othertoolsservice/2016/06/09/3d-snfg-list-of-residue-names/index.html

In addition of using the GLYCAM Webserver to obtain the known Sugar abbreviations and also
included the aglycans that I obtained from the GLYCAM Weberserver.

The Pytest Files were retrieved from RCSB-PDB and the GLYCAM-Webserver:

https://glycam.org/

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

## Developers certificate of origin
- [x] I certify that this contribution is covered by the LGPLv2.1+ license as defined in our [LICENSE](https://github.com/MDAnalysis/mdanalysis/blob/develop/LICENSE) and adheres to the [**Developer Certificate of Origin**](https://developercertificate.org/).


<!-- readthedocs-preview mdanalysis start -->
----
📚 Documentation preview 📚: https://mdanalysis--4790.org.readthedocs.build/en/4790/

<!-- readthedocs-preview mdanalysis end -->

talagayev committed:
Update selection.py

Implementation of SugarSelection with the known abbreviations and aglycans obtained from the glycam webserver

pep8speaks commented:
Hello @talagayev! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:







There are currently no PEP 8 issues detected in this Pull Request. Cheers! :beers: 

##### Comment last updated at 2025-01-11 17:54:06 UTC

talagayev committed:
Update selection.py

pep fix 1

talagayev committed:
Update selection.py

talagayev committed:
Update selection.py

talagayev committed:
Added Test Files for GLYCAM and PDB sugars

talagayev committed:
Update datafiles.py

Added test files

talagayev committed:
Update test_atomselections.py

Added tests for sugar selection

talagayev committed:
Update test_atomselections.py

pep fix

talagayev committed:
Update datafiles.py

talagayev commented:
@hmacdope I will ping you here since you were one of the participants in the discussion that initiated the Issue that this PR is covering.

The main problem that I see currently, would be that the GLYCAM Abbreviations for the sugars due to the combinations have quite a lot of different names/abbreviations.

Some of them could lead to tricky cases, with the Allose Nomenclature having `RNA` as one of the abbreviations. 

I checked the RCSB-PDB and found only one case, where a unique ligand was called `RNA`, but still could be dangerous if the 
users call something "RNA" in their files.

As for the coverage, while looking at the PDBs in RCSB-PDB it covers `NAG`, `GLC` etc., which would be convenient if somebody wants to select those in the PDB Files, but does not cover for example `Glycerol`, which makes sense since it is not a sugar, but is quite common in crystal structures among those sugars as `NAG`, `GLC`. Does it make sense to have a selection that would somehow covers both cases of Glycerol and similar compounds together with sugars?

talagayev committed:
Merge branch 'develop' into sugar_selection

talagayev committed:
Update CHANGELOG

added changelog

talagayev committed:
Update selection.py

added docs

RMeli reviewed (approved):
LGTM

IAlibay commented on package/MDAnalysis/core/selection.py:
```diff
@@ -1056,6 +1056,258 @@ def _apply(self, group):
+      * manually added aglycans 'ROH', 'OME', 'TBT'
+        from GLYCAM-Web generated files.
+
+    .. versionadded:: 2.9.0
```
```suggestion
    .. versionadded:: 2.11.0
```

IAlibay commented on package/MDAnalysis/core/selection.py:
```diff
@@ -1056,6 +1056,258 @@ def _apply(self, group):
+        return group[np.isin(nmidx, matches)]
+
+
+
```
```suggestion
```

IAlibay commented on testsuite/MDAnalysisTests/data/6kya.pdb:
gzip this file?

talagayev replied:
compressed the file to `6kya.pdb.gz`

IAlibay commented on testsuite/MDAnalysisTests/data/GLYCAM_sugars.pdb:
This is small enough to be ok non-gzipped.

IAlibay commented on testsuite/MDAnalysisTests/datafiles.py:
```diff
@@ -879,5 +881,8 @@
 # DSSP testing: from https://github.com/ShintaroMinami/PyDSSP
 DSSP = (_data_ref / "dssp").as_posix()
 
+GLYCAM = (_data_ref / 'GLYCAM_sugars.pdb').as_posix()
```
Add some comments to describe these files?

talagayev replied:
Added the description :)

IAlibay reviewed (changes requested):
Needs updating against main and a couple of small things, otherwise this is good.

talagayev committed:
Merge branch 'develop' into sugar_selection

IAlibay commented on package/CHANGELOG:
```diff
@@ -81,6 +81,7 @@ Fixes
    directly passing them. (Issue #3520, PR #5006)
 
 Enhancements
+ * Addition of 'sugar' token for GLYCAM, PDB and CHARMM sugar selection (Issue #4790)
```
Sorry my bad - it should go in 2.11! (I forget what we have and haven't released)

talagayev replied:
All good, will adjust it :)

talagayev replied:
adjusted it t 2.11

talagayev commented:
I also wanted to get your quick opinion on that PR @IAlibay, since when we implemented the `water` token in one the PRs you mentioned that there can be the case, that making the tokens will lead to users not carefully checking their files and thus deleting something, so that while it improved the user-friendliness it also led to a potential increase in user mistakes.

Here I have a small worry with this PR, that with the big amount of possible of abbrevations with 3 letters some may overlap with co-crystalized ligands. So for example something like `NAG` I only saw for the sugars in PDBs, so I would be suprised if a ligand would have that abbreviation, but something like `6lA` or in that direction I would not be suprised to encounter in some PDB as the identifier for the co-crystalized ligand and thus may lead to the co-crystalized ligand being selected as soon as the user uses the `sugar` token.

What would be your opinion on that, how likely something like that could happen, espeically again with here the token covering every possible GLYCAM nomenclature, which leads to many ids falling under it 🤔

talagayev committed:
Update package/MDAnalysis/core/selection.py

Co-authored-by: Irfan Alibay <IAlibay@users.noreply.github.com>

talagayev committed:
Update CHANGELOG

moved to 2.11.0

talagayev committed:
Update package/MDAnalysis/core/selection.py

Co-authored-by: Irfan Alibay <IAlibay@users.noreply.github.com>

talagayev committed:
compression of 6kya.pdb and addition of note about files in datafiles.py

talagayev committed:
black formatting

talagayev committed:
black formatting

IAlibay committed:
Merge branch 'develop' into sugar_selection

IAlibay commented:
@talagayev the number of tokens being added is rather large and indeed the increasing risk of collisions / "too much hand-holding" is possible.

I'm not 100% sure about how much of an issue this really is. @orbeckst I know you have been more involved with the selection keywords - do you have any thoughts? Is this too large a range of residues we are trying to cover?

talagayev commented:
> @talagayev the number of tokens being added is rather large and indeed the increasing risk of collisions / "too much hand-holding" is possible.
> 
> I'm not 100% sure about how much of an issue this really is. @orbeckst I know you have been more involved with the selection keywords - do you have any thoughts? Is this too large a range of residues we are trying to cover?

I think it will be a very rare case, but that one would fall under such a case with the ligand having the name of an sugar token.

https://www.rcsb.org/structure/4G3G

but overall maybe we can make a warning when we use that selection something among the lines of "Certain topology files contain ligands with abbreviations used in the sugar token selection and may be affected" so that we then state that even if that is rare case we still inform the user about it and then it is up to him to double check that 🤔

orbeckst commented on package/MDAnalysis/core/selection.py:
```diff
@@ -1206,6 +1206,1646 @@ def _apply(self, group):
+        "QgB",
+        "PgA",
+        "PgB",
+        # N-Acetyl Glucosamine Nomenculature
```
Nomenclature

orbeckst commented on package/MDAnalysis/core/selection.py:
```diff
@@ -1206,6 +1206,1646 @@ def _apply(self, group):
+        "UyB",
+        "QyA",
+        "QyB",
+        # Glucosamine Nomenculature
```
Nomenclature

orbeckst commented on package/MDAnalysis/core/selection.py:
```diff
@@ -1206,6 +1206,1646 @@ def _apply(self, group):
+        "3ys",
+        "4yS",
+        "4ys",
+        # Glucuronic Acid Nomenculature
```
Nomenclature

orbeckst commented on package/MDAnalysis/core/selection.py:
```diff
@@ -1206,6 +1206,1646 @@ def _apply(self, group):
+        "TzA",
+        "TzB",
+        "0ZBP",
+        # Quinovose Nomenculature
```
Nomenclature

orbeckst commented on package/MDAnalysis/core/selection.py:
```diff
@@ -1206,6 +1206,1646 @@ def _apply(self, group):
+
+    sugar_res = {
+        # https://glycam.org/docs/othertoolsservice/2016/06/09/3d-snfg-list-of-residue-names/index.html
+        # Full PDB Abbreviations Nomenclature
```
Nomenclature

orbeckst replied:
please correct spelling throughout in comments

talagayev replied:
Ah my bad, didn't notice that typo at all, will fix it.

orbeckst commented on package/MDAnalysis/core/selection.py:
```diff
@@ -1206,6 +1206,1646 @@ def _apply(self, group):
+    sugar_res = {
+        # https://glycam.org/docs/othertoolsservice/2016/06/09/3d-snfg-list-of-residue-names/index.html
+        # Full PDB Abbreviations Nomenclature
+        "GLC",
```
Is this really how black wants to format it???? Ugly... but ok.

talagayev replied:
Yes black formatting is tricky with that many tokens, no idea comes to my mind how to make it more clean 🤔

orbeckst replied:
It’s ok — I was just sighing out loudly ;-).

orbeckst replied:
Ok, I officially hate this formatting because it takes forever to scroll through.

Can we manually format it just across our linewidth and tell black to not touch it??

orbeckst commented on package/MDAnalysis/core/selection.py:
```diff
@@ -1206,6 +1206,1646 @@ def _apply(self, group):
+        nmidx = resname_attr.nmidx[group.resindices]
+        # intersect atom's resname index and matches to sugar_res
+        return group[np.isin(nmidx, matches)]
+
```
remove the comments – we don't have them for the identical code for the other named selections.

orbeckst commented on package/MDAnalysis/core/selection.py:
```diff
@@ -1206,6 +1206,1646 @@ def _apply(self, group):
+        "TBT",
+    }
+
+    def _apply(self, group):
```
This whole method is effectively copied & pasted multiple times (for water, for nucleic acids, ...) and the only difference is the line
```python
if nm in self.sugar_res
```

What do we think about adding a helper method to the base class  so that we can use code along the lines of
```python
def _apply(self, group):
    return self._apply_match_by_resnames(self.sugar_res)
```
and reducing overall code footprint?

talagayev replied:
Yes, was some time ago when I wrote the inital code, but yes it was mainly basically just reusing the already existing style of selection that was in there and basically just reusing it for sugars. I think a similar approach was done when I coded the `water` selection token, but there it was less tokens so it worked well for that one.

@orbeckst For the helper function, agree would be good to refactor a little bit the code, since currently it is `_apply` with small modifications for every selection token. Do we want to start it in that PR and then adjust to the other ones or merge that and create a new Issue with the refactoring and refactor?

orbeckst replied:
When you put it like this then doing a separate PR for the refactor is better: that’s an easy “lgtm”. Would you mind doing that and then coming back here?

talagayev replied:
yup will do. will create then an Issue suggesting the refactoring, look into the new bacterial glycans and adress the comments and then ping you / rerequest review. Getting a little bit late, will be able to do that tomorrow during the day :)

orbeckst reviewed (changes requested):
This is a huge number of specific residue names. The danger is that it becomes outdated and when people rely on "sugar" catching everything, they miss something. Perhaps the better place would be a guesser where people then have to demonstrate intention for having a convenience selection (although I don't even know if our guessers would currently come with special selections).

EDIT: Collisions — yes, that's annoying, like `CA` for C-alpha and `CA` for Ca<sup>2+</sup>. I think we can make a choice, add a warning to the docs along what @talagayev suggests but then the main thing to do would be to give the user the information _which_ resnames are considered sugar, i.e., have the full list appear in the docs. (Similar for "protein", "nucleic acids" etc. One way to do this would be to store them in variables and document the variables so that the contents show up in the generated docs.

EDIT 2: And document in the Userguide, eg https://userguide.mdanalysis.org/stable/standard_selections.html#protein-selection

On the other hand, I kind of doubt that these lists change rapidly over time, so overall I am tending towards being ok with embedding them into the code.

I am a bit concerned with the fact that we're essentially using copy & paste for the `_apply()` code. I'd rather break that pattern now and refactor it into a helper method, e.g., `Selection._apply_mathc_by_resname()`, and use it for _all_ of these named selections. Or is there a good reason for not doing that?

talagayev commented:
> On the other hand, I kind of doubt that these lists change rapidly over time, so overall I am tending towards being ok with embedding them into the code.

I checked quickly and usually they update them rarely, so the last entry I saw was 2016, the one that was used for getting the abbreviations that we use here, but they have a recent preprint out for bacterial glycans. quite a new one:

https://doi.org/10.64898/2026.01.26.700271

I will look into it to see if we have covered those or if those are new and need to be added as well :)

orbeckst commented:
Great, thanks. Am 2/21/26 um 13:40 schrieb Valerij Talagayev ***@***.***>:﻿
@talagayev commented on this pull request.



In package/MDAnalysis/core/selection.py:
> +        "CgL",
+        "DgL",
+        "EgL",
+        "FgL",
+        "GgL",
+        "HgL",
+        "IgL",
+        "JgL",
+        "KgL",
+        # Aglycon Nomenculature
+        "ROH",
+        "OME",
+        "TBT",
+    }
+
+    def _apply(self, group):

yup will do. will create then an Issue suggesting the refactoring, look into the new bacterial glycans and adress the comments and then ping you / rerequest review. Getting a little bit late, will be able to do that tomorrow during the day :)

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

talagayev committed:
Merge branch 'develop' into sugar_selection

talagayev committed:
fixed Nomenclature typo

talagayev committed:
removed comments

talagayev committed:
added GLYCAM Bacterial Nomenclature

talagayev commented:
@orbeckst I adjusted the typo error, removed the comments as discussed and opened up #5247 to create an Issue for the refactoring of the code to not reuse `__apply()` in each individual selection class.

orbeckst commented:
PR #5294 refactored the residue naming logic.

orbeckst committed:
Merge branch 'develop' into sugar_selection

orbeckst commented on package/MDAnalysis/core/selection.py:
```diff
@@ -1209,6 +1209,1661 @@ def _apply(self, group):
+        ]
+        nmidx = resname_attr.nmidx[group.resindices]
+
+        return group[np.isin(nmidx, matches)]
```
```suggestion
        return self._apply_match_by_resnames(group, self.sugar_res)
```

orbeckst replied:
Makes use of the refactor from PR #5294

orbeckst reviewed (approved):
I'll make a change to use the refactored resname matching and then it's ok from my end.

Can you please also check that the resnames show up in the documentation **and** raise an issue in the User Guide to update the selections section there https://userguide.mdanalysis.org/stable/standard_selections.html ?

(If you can solve the black-formatting issue then that would be a bonus but it's not a blocker.)

orbeckst committed:
use _apply_match_by_resnames
